linter/lints/crl/lints.go

```diff
@@ -0,0 +1,381 @@
+		"noZeroReasonCodes":              noZeroReasonCodes,
+		"hasNoCertIssuers":               hasNoCertIssuers,
+		"hasAcceptableValidity":          hasAcceptableValidity,
+		"noCriticalReasons":              noCriticalReasons,
+		"noCertificateHolds":             noCertificateHolds,
+		"hasMozReasonCodes":              hasMozReasonCodes,
+	}
+}
+
+// getExtWithOID is a helper for several lints in this file. It returns the
+// extension with the given OID if it exists, or nil otherwise.
+func getExtWithOID(exts []pkix.Extension, oid asn1.ObjectIdentifier) *pkix.Extension {
+	for _, ext := range exts {
+		if ext.Id.Equal(oid) {
+			return &ext
+		}
+	}
+	return nil
+}
+
+// LintCRL examines the given lint CRL, runs it through all of our checks, and
+// returns a list of all failures
+func LintCRL(lintCRL *crl_x509.RevocationList) *zlint.ResultSet {
+	rset := zlint.ResultSet{
+		Version:   0,
+		Timestamp: time.Now().UnixNano(),
+		Results:   make(map[string]*lint.LintResult),
+	}
+
+	type namedResult struct {
+		Name   string
+		Result *lint.LintResult
+	}
+	resChan := make(chan namedResult, len(registry))
+
+	for name, callable := range registry {
+		go func(name string, callable crlLint) {
+			resChan <- namedResult{name, callable(lintCRL)}
+		}(name, callable)
+	}
+
+	for i := 0; i < len(registry); i++ {
+		res := <-resChan
+		switch res.Result.Status {
+		case lint.Notice:
+			rset.NoticesPresent = true
+		case lint.Warn:
+			rset.WarningsPresent = true
+		case lint.Error:
+			rset.ErrorsPresent = true
+		case lint.Fatal:
+			rset.FatalsPresent = true
+		}
+		rset.Results[res.Name] = res.Result
+	}
+
+	return &rset
+}
+
+// hasIssuerName checks RFC 5280, Section 5.1.2.3:
+// The issuer field MUST contain a non-empty X.500 distinguished name (DN).
+// This lint does not enforce that the issuer field complies with the rest of
+// the encoding rules of a certificate issuer name, because it (perhaps wrongly)
+// assumes that those were checked when the issuer was itself issued, and on all
+// certificates issued by this CRL issuer. Also because there are just a lot of
+// things to check there, and zlint doesn't expose a public helper for it.
+func hasIssuerName(crl *crl_x509.RevocationList) *lint.LintResult {
+	if len(crl.Issuer.Names) == 0 {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: "CRLs MUST have a non-empty issuer field",
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+// usesCorrectTimeEncoding checks RFC 5280, Section 5.1.2.4 and 5.1.2.5:
+// CRL issuers conforming to this profile MUST encode thisUpdate and nextUpdate
+// as UTCTime for dates through the year 2049. UTCTime and GeneralizedTime
+// values MUST be expressed in Greenwich Mean Time (Zulu) and MUST include
+// seconds, even where the number of seconds is zero.
+func usesCorrectTimeEncoding(crl *crl_x509.RevocationList) *lint.LintResult {
+	// TODO: Figure out how to do this gracefully. Zlint's three lints for this
```

Comment:
File an issue for this? Maybe it would be better to not have this function at all, so as not to give the impression it's implemented.